### ISSUE TITLE
IOS Resolution is horribly slow and happens on every save

### ISSUE THREAD

**Issue by tommikiviniemi-srs** [googlesamples/unity-jar-resolver#78]
Hi,

This is a new problem which I think is caused by the iOS resolver always enumerating the project for XML files on every save, as it didn't happen with an older version of the resolver libraries.

We only run the resolvers/downloaders to fetch libraries to the project, which means updating libraries is a controlled action. However, one of the more recent resolver updates stalls the Unity editor for around 30-60 seconds every time a source file is saved, if the current deployment target is set to iOS. I checked the sources and the resolver does A LOT of work *even when the iOS resolver is disabled* (we programmatically disable the resolvers et al whenever sources are saved to enforce no additional wait time when sources are saved and re-enable the appropriate parts at build time, this used to work before).

Could you please change the sources to only run the relevant actions in the [InitializeOnLoad] functions *only* when the appropriate resolvers etc. are enabled? I will now have to resort having to hack in this functionality because this really hampers our production. This problem largely goes away if the iOS resolver dll is deleted, the other resolver dlls also suffer of this but to a lesser degree and are not as critical to fix.

Thanks!
- Tommi

**Comment by tommikiviniemi-srs** [comment]
I updated the resolver again, this time from 1.2.52.0 to 1.2.54.0 and the performance problem seems to be largely gone. I didn't see anything related to this in the release notes but it's almost acceptable now. However, it's still quite a lot faster with the iOS resolver completely removed from the project, so is there any chance you could go through all the DLLs and honour the enabled status not to do any work unless needed?

Thanks again,
- Tommi

**Comment by stewartmiles** [comment]
Hi @tommikiviniemi-srs 

We sped up the dependency checking process in https://github.com/googlesamples/unity-jar-resolver/issues/70 but alas, it sounds like the asset database is still pretty mind bogglingly slow.

Since the iOS resolver only runs as a post-build step I'll change this to only load dependencies from XML files into the post-build / project export process for iOS instead.  That way you don't need to disable the resolver for the editor and enable to build.

Cheers,
Stewart

**Comment by tommikiviniemi-srs** [comment]
Hi Stewart and thanks for the reply!

That sounds really good, it'd be awesome if as little as possible was done when saving sources in case the corresponding functionality is not active. I guess this is one of the things that show up only when you have an absolutely massive project, but its impact is quite large on the team as saving sources is something that's done pretty frequently.

You said the iOS resolver would only be run as a post-build step, shouldn't it also run when manually triggering an update to the libraries or manually resolving from the menu? We actually prefer to, when possible, to let the programmers run the resolvers at a time when we have scheduled QA time to update the native libraries, and then check in the libraries to our SCM.

Thanks again,
- Tommi

**Comment by stewartmiles** [comment]
@tommikiviniemi-srs regarding the iOS resolver, at the moment it always runs as a post-build step it doesn't allow manual resolution at the moment I'm afraid.  We have an issue open to support manual resolution https://github.com/googlesamples/unity-jar-resolver/issues/59 but it's way more complicated than the Android side.

In terms of making sure things use the same libraries.  We recommend folks specify exact dependencies for Cocoapods in their *Dependencies.xml files.  Over in Firebase we pin each Unity plugin release to a specific version of the required iOS pods which means your build won't magically update over time without you knowing about it.  If you want to be really sure and keep everything local, you could do what we do internally with our build systems and setup a local Cocoapods repo that only contains what you need, it's work but possible.

Cheers,
Stewart

**Comment by tommikiviniemi-srs** [comment]
Ah you are right Stewart, I was thinking of Android for the manual resolution step.

In fact we wanted to use the resolver for iOS Cocoapods in Workspace integration mode for one of our plugins (Urbanairship), but it was held up by https://github.com/googlesamples/unity-jar-resolver/issues/71 so it's disabled altogether now in iOS mode. But that's another issue.

I see you already committed a fix for this so thank you so much for that, really appreciate the support!

Thanks,
- Tommi

**Comment by johnb003** [comment]
In the plugin released with Firebase 4.2.1 we've updated this resolution to only happen during build.
